### PR TITLE
caching table-in-out-functions & chunk cache refactor

### PR DESCRIPTION
**Pull request by samansmink**
This PR:
- Adds generalized api for PhysicalOperators to cache output.
- Refactored the existing chunk cache mechanism (Implemented as CachingPhysicalOperator base class)
- Chunk cache now also works for streaming results
- Adds caching TableFunction in-out-functions

The motivation behind this change is to be able to implement a caching in-out-function to be able to stream a query result as serialized blobs. 

Currently, duckdb has a chunk cache mechanism which is used prevent to low cardinality vectors from being pushed through the pipeline, for example after a filter with low selectivity.

This pr adds a more generalized version of such a caching mechanism. It works by adding a `bool RequiresFinalExecute()` method to the PhysicalOperator class which allows an operator to indicate that it may hold some cached data even though it's emitting an empty result. Note that this is different from emitting OperatorResultType::HAVE_MORE_OUTPUT, because that will cause the operator to be called again using the same input. To ensure that operators holding cached results properly flush their output after input runs out, the PipelineExecutor will call the PhysiscalOperator 1 last time after the whole pipeline is processed:
```
void PhysicalOperator::FinalExecute(ExecutionContext &context, DataChunk &chunk, GlobalOperatorState &gstate, OperatorState &state);
```
Note that there the operator can only output data here, there is no input chunk. Also, the operator can not return an OperatorResultType because this is the last time this operator will be called. The idea here is that operators can never cache more than 1 chunk (because they would've emitted it during normal execution).

### CODE REVIEW

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. Some comments:

File: src/execution/physical_operator.cpp [hunk 1]
```diff
@@ -51,6 +51,11 @@ OperatorResultType PhysicalOperator::Execute(ExecutionContext &context, DataChun
                                              GlobalOperatorState &gstate, OperatorState &state) const {
 	throw InternalException("Calling Execute on a node that is not an operator!");
 }
+
+void PhysicalOperator::FinalExecute(ExecutionContext &context, DataChunk &chunk, GlobalOperatorState &gstate,
+                                    OperatorState &state) const {
+	throw InternalException("Calling FinalExecute on a node that is not an operator!");
```
[the remark this answers is not in the document]
Comment (samansmink) [reply src/execution/physical_operator.cpp:57]: yea makes perfect sense! will do

File: src/function/table/summary.cpp [hunk 2]
```diff
@@ -22,7 +22,7 @@ static unique_ptr<FunctionData> SummaryFunctionBind(ClientContext &context, Tabl
 }
 
 static OperatorResultType SummaryFunction(ExecutionContext &context, TableFunctionInput &data_p, DataChunk &input,
-                                          DataChunk &output) {
+                                          DataChunk &output, bool enable_caching) {
```
[the remark this answers is not in the document]
Comment (samansmink) [reply src/function/table/summary.cpp:25]: I think this will indeed be irrelevant after processing you other comment on the cache disabling

**Comment by samansmink** [comment]
@Mytherin CI failures seem to be unrelated, cannot reproduce the regressions.

**Review by Mytherin — COMMENTED** [review]
Thanks for the updates! A few more comments:

**Comment by samansmink** [comment]
@Mytherin fixed all comments! Could indeed do without the tmp_chunk, now using the final_chunk and intermediate chunks

**Comment by Mytherin** [comment]
Thanks! LGTM

**Pull request merged**